File: web/src/features/integrity/timeline.ts
import type { IntegrityEvent } from "@/features/attempts/api";

/** The G-05 filter chips. */
export type TimelineFilter = "all" | "away" | "audio" | "network";

const AWAY_OPENERS = new Set(["tab_hidden", "window_blur", "fullscreen_exit"]);
const AWAY_CLOSERS = new Set(["tab_visible", "window_focus", "fullscreen_enter"]);
const AUDIO = new Set(["audio_play", "audio_ended", "audio_blocked"]);
const NETWORK = new Set(["network_offline", "network_online"]);
const CLOSERS = new Set([...AWAY_CLOSERS, "network_online", "audio_ended"]);

/** Which chip an event belongs under; `null` for kinds no chip narrows to. */
export function family(kind: string): Exclude<TimelineFilter, "all"> | null {
  if (AWAY_OPENERS.has(kind) || AWAY_CLOSERS.has(kind)) return "away";
  if (AUDIO.has(kind)) return "audio";
  if (NETWORK.has(kind)) return "network";
  return null;
}

/** One row of the Diễn biến table. */
export interface TimelineRow {
  event: IntegrityEvent;
  /** No return yet: the student is still away, still offline, still listening. */
  ongoing: boolean;
  /** For `audio_play`, which play of that question this was. */
  playNo: number | null;
}

/**
 * The rows the table draws. The server paired every episode onto its opening
 * event, so a closing event carries nothing the row above does not; and a
 * second leave inside an open episode (blur then hidden) is the same absence
 * twice. Both are dropped. An opener with no duration is kept only when nothing
 * later closed its family -- the "đang tiếp diễn" row (G-05b).
 */
export function timelineRows(
  events: IntegrityEvent[],
  filter: TimelineFilter,
): TimelineRow[] {
  const plays = new Map<string, number>();
  const rows: TimelineRow[] = [];
  events.forEach((event, index) => {
    if (CLOSERS.has(event.kind)) return;
    let playNo: number | null = null;
    if (event.kind === "audio_play") {
      const key = event.questionId ?? "";
      playNo = (plays.get(key) ?? 0) + 1;
      plays.set(key, playNo);
    }
    const opener =
      AWAY_OPENERS.has(event.kind) ||
      event.kind === "network_offline" ||
      event.kind === "audio_play";
    const ongoing = opener && event.durationMs == null && !closedLater(events, index);
    if (opener && event.durationMs == null && !ongoing) return;
    if (filter !== "all" && family(event.kind) !== filter) return;
    rows.push({ event, ongoing, playNo });
  });
  return rows;
}

function closedLater(events: IntegrityEvent[], index: number): boolean {
  const opener = events[index];
  if (opener === undefined) return false;
  const fam = family(opener.kind);
  for (let i = index + 1; i < events.length; i++) {
    const later = events[i];
    if (later === undefined) continue;
    if (!CLOSERS.has(later.kind)) continue;
    if (fam === "audio" && later.questionId !== opener.questionId) continue;
    if (family(later.kind) === fam) return true;
  }
  return false;
}

/** The trailing episode the strip must say it is not summing (G-05b). */
export function hasOpenEpisode(events: IntegrityEvent[]): boolean {
  return timelineRows(events, "away").some((row) => row.ongoing);
}

/** "2:41" from milliseconds, the strip's and the table's shape. */
export function clockSpan(ms: number): string {
  const total = Math.max(0, Math.round(ms / 1000));
  const h = Math.floor(total / 3600);
  const m = Math.floor((total % 3600) / 60);
  const s = total % 60;
  const pad = (n: number) => String(n).padStart(2, "0");
  return h > 0 ? `${h}:${pad(m)}:${pad(s)}` : `${m}:${pad(s)}`;
}

File: server/internal/integrity/timeline.go
// Package integrity turns an attempt's event log into §10.4's timeline: paired
// durations, offsets from the start, and the summary strip. Counts, never
// verdicts -- the teacher judges; the app reports.
package integrity

import (
	"sort"
	"time"
)

// Event is one row of the log as the teacher sees it.
type Event struct {
	ID         int64
	Kind       string
	OccurredAt time.Time
	ReceivedAt time.Time
	// ClientSeq is nil for a server-written event (`resume`, `session_takeover`).
	ClientSeq  *int
	SessionID  string
	QuestionID *string
	Meta       []byte

	// OffsetMs is milliseconds since the attempt started, never negative.
	OffsetMs int
	// DurationMs is set on the event that opened a paired episode; nil on one
	// that never closed (§10.4).
	DurationMs *int
}

// Summary is the strip above the list.
type Summary struct {
	TotalAwayMs     int
	AwayEpisodes    int
	PasteCount      int
	ResumeCount     int
	AudioReplays    int
	OfflineEpisodes int
}

type Timeline struct {
	StartedAt time.Time
	Events    []Event
	Summary   Summary
}

// Away episodes follow the engine's own rule (useIntegrityMonitor): the first
// leave opens one, the first return closes it, whichever of the two signals
// fires first. Counting blur and hidden separately would report two absences
// for one alt-tab.
var (
	leaves  = map[string]bool{"tab_hidden": true, "window_blur": true}
	returns = map[string]bool{"tab_visible": true, "window_focus": true}
	// pairs are the other paired kinds, each closed by exactly one other kind.
	pairs = map[string]string{
		"fullscreen_exit": "fullscreen_enter",
		"network_offline": "network_online",
		"audio_play":      "audio_ended",
	}
)

// Build orders the log, pairs what pairs, and totals the strip.
//
// minAwayMs is the assignment's threshold: an episode shorter than it is not
// counted, exactly as the engine did not count it against the student.
// audioReplays comes from the plays table rather than the log, because the
// count that mattered was the server's (§11.4).
func Build(startedAt time.Time, minAwayMs, audioReplays int, events []Event, now time.Time) Timeline {
	order(events)
	for i := range events {
		events[i].OffsetMs = max(0, int(events[i].OccurredAt.Sub(startedAt)/time.Millisecond))
		events[i].DurationMs = nil
	}
	summary := tally(events, minAwayMs, now)
	summary.AudioReplays = audioReplays
	pairOthers(events)
	return Timeline{StartedAt: startedAt, Events: events, Summary: summary}
}

// tally pairs each leave with the next return, writing the duration onto the
// leave, and counts the rest of the strip as it goes.
func tally(events []Event, minAwayMs int, now time.Time) Summary {
	var summary Summary
	open := -1
	for i := range events {
		e := &events[i]
		switch {
		case leaves[e.Kind]:
			if open < 0 {
				open = i
			}
		case returns[e.Kind]:
			if open < 0 {
				continue
			}
			d := span(events[open].OccurredAt, e.OccurredAt)
			events[open].DurationMs = &d
			summary.TotalAwayMs += d
			if d >= minAwayMs {
				summary.AwayEpisodes++
			}
			open = -1
		case e.Kind == "paste":
			summary.PasteCount++
		case e.Kind == "resume":
			summary.ResumeCount++
		case e.Kind == "network_offline":
			summary.OfflineEpisodes++
		}
	}
	// Still away: counted, not summed (G-05b).
	if open >= 0 && span(events[open].OccurredAt, now) >= minAwayMs {
		summary.AwayEpisodes++
	}
	return summary
}

// pairOthers closes each opener with the next closer of its kind -- per
// question for audio, since two players can be open at once.
func pairOthers(events []Event) {
	open := map[string]int{}
	for i := range events {
		e := &events[i]
		if closer, ok := pairs[e.Kind]; ok {
			open[key(closer, e.QuestionID)] = i
			continue
		}
		if j, ok := open[key(e.Kind, e.QuestionID)]; ok {
			d := span(events[j].OccurredAt, e.OccurredAt)
			events[j].DurationMs = &d
			delete(open, key(e.Kind, e.QuestionID))
		}
	}
}

func key(kind string, questionID *string) string {
	if questionID == nil {
		return kind
	}
	return kind + ":" + *questionID
}

func span(from, to time.Time) int {
	return max(0, int(to.Sub(from)/time.Millisecond))
}

// order sorts sessions by when they began and, within a session, by
// `client_seq` -- so a skewed client clock cannot scramble a session's own
// order (§10.6). The server's `resume` opens a session and `session_takeover`
// closes one, so they sit at the ends.
func order(events []Event) {
	began := map[string]time.Time{}
	for _, e := range events {
		if first, ok := began[e.SessionID]; !ok || e.ReceivedAt.Before(first) {
			began[e.SessionID] = e.ReceivedAt
		}
	}
	sort.SliceStable(events, func(i, j int) bool { return before(events[i], events[j], began) })
}

// before is order()'s comparator: session start, then the server's place for
// resume/takeover, then the client's own sequence, then the wall clock.
func before(a, b Event, began map[string]time.Time) bool {
	if a.SessionID != b.SessionID {
		if !began[a.SessionID].Equal(began[b.SessionID]) {
			return began[a.SessionID].Before(began[b.SessionID])
		}
		return a.SessionID < b.SessionID
	}
	if ra, rb := place(a), place(b); ra != rb {
		return ra < rb
	}
	if a.ClientSeq != nil && b.ClientSeq != nil && *a.ClientSeq != *b.ClientSeq {
		return *a.ClientSeq < *b.ClientSeq
	}
	if !a.OccurredAt.Equal(b.OccurredAt) {
		return a.OccurredAt.Before(b.OccurredAt)
	}
	return a.ID < b.ID
}

func place(e Event) int {
	switch e.Kind {
	case "resume":
		return 0
	case "session_takeover":
		return 2
	default:
		return 1
	}
}
